File: 20. Dynamic Programming/partition_equal_subset_sum.cpp
class Solution {
public:
    bool subsetSum(vector<int>&nums, int halfSum){
        //lets create dp table and inititalize it 
        
        //lets create n and w 
        int n=nums.size(), w=halfSum;
        
        //make a dp table and initialize it 
        bool dp[n+1][w+1];
        for(int i=0;i<n+1;i++){
            for(int j=0;j<w+1;j++){
                if(i==0) dp[i][j]=false;
                if(j==0) dp[i][j]=true;
            }
        }
        
        //now lets write our subset sum code ie. a variation of 0/1 knapsack
        for(int i=1;i<n+1;i++){
            for(int j=1;j<w+1;j++){
                if(nums[i-1]<=j)
                    dp[i][j] = dp[i-1][j-nums[i-1]] || dp[i-1][j];
                else 
                    dp[i][j] = dp[i-1][j];
            }
        }
        
        return dp[n][w];
        
    }
    
    bool canPartition(vector<int>& nums) {
        int sum=0;
        
        for(int i=0;i<nums.size();i++){
            sum+=nums[i];
        }
        
        if(sum%2!=0){
            //if the sum is odd then can never divide the array into two equal subsets
            return false;
        } 
        
        else {
            //now this is a problem which I have studied previously 
            //if we find a subset whose addition is sum/2 
            //then we just have to call subset sum function which tells us that that subset sun of sum/2
            //exists or not 
            //if it exists then we just have to return ture else false
            //hence lets make a subset sum function
            int halfSum=sum/2;
            return subsetSum(nums, halfSum);
        }
    }
};

File: 20. Dynamic Programming/PartitionEqualSubsetSum.java
public class PartitionEqualSubsetSum{

   public static void main(String args[]){
     int nums[]={1,3,5,2,6,3};
     System.out.print(canPartition(nums));
   }
    public static boolean canPartition(int[] nums) {
        int n=nums.length;
        int sum=0;
        for(int i=0;i<n;i++){
            sum+=nums[i];
        }
        if(sum%2!=0){
            return false;
        }
        
        int dp[][]=new int[nums.length+1][(sum/2)+1];
        for(int i=0;i<dp.length;i++){
            for(int j=0;j<dp[0].length;j++){
                dp[i][j]=-1;
            }
        }

        return subset(nums,sum/2,0,dp)==1;
    } 
    static int subset(int arr[],int val,int ind,int dp[][]){
        if(ind>=arr.length){
            if(val==0) return 1 ;
            return 0;
        }
        if(dp[ind][val]!=-1){
           return dp[ind][val];
        }
        int taken=0;
        if(arr[ind]<=val){
        taken=subset(arr,val-arr[ind],ind+1,dp);
        }
        int nottaken=subset(arr,val,ind+1,dp);
       return dp[ind][val]=(taken | nottaken);
       
        }
        
}
